typing/typecore.ml

```diff
@@ -778,21 +791,66 @@ let solve_constructor_annotation
         | _ -> assert false
   in
   if ids <> [] then ignore begin
+    let get_decl_manifest id env =
+      let decl =
+        try Env.find_type (Path.Pident id) env with Not_found -> assert false in
+      match decl.type_manifest with
+        None -> assert false
+      | Some ty -> (decl, ty)
+    in
     let ids = List.map (fun x -> x.txt) ids in
     let rem =
+      (* First process the existentials introduced by this constructor.
+         Just need to make their definitions abstract. *)
```

Comment:
Naive question: why do we need to do two passes, one on the constructor existential instances and another on the bound existential names? Couldn't we do a single pass on the bound existential names, with logic that would also work well on the purely existential instances?

Reply:
While this would be possible, this would be less robust. Namely, for internal existentials we delay the creation of the abstract type, so that we will never show a strange internal name for the user. For existentials that were introduced by unification, we cannot do that, so we just create alias names, hoping that the strange internal name will not show up later.